### PR TITLE
batches: improve performance of viewerBatchChangesCodeHosts

### LINKED ISSUE #37246: Batch changes and specs with large numbers of changesets and diffstats can be slow
I started investigating this last week, but I need to break off for now and come back to this later.

Basically, executing a batch spec that creates, say, 200+ changesets, each with diffstats around 200+ lines added, results in the GraphQL queries that drive the preview and batch change pages being slow, in the order of multiple seconds. This appears to correspond to increased database load.

Poking around with [a tool to isolate each field in a GraphQL query](https://github.com/LawnGnome/graphql-field-timer) (self promotion alert), the slowness mostly seems to come from two areas:

1. Retrieving fields related to the `codeHost`. This is a fairly uncomfortable abstraction in Go over the lower level external service and repo tables, and is presumably missing some indices and/or retrieving more data than we need to answer the real questions we need to answer on these pages (which is basically "do you have webhooks enabled?").
2. Calculating the total diffstat for the batch spec. This happens in the GraphQL layer by constructing _all_ the changeset spec resolvers for the batch spec, then adding their diffstats together, presumably for permission reasons (since users should be unaware of diffstats for hidden changesets). We need to find a way to move that into the database if possible, I think. (Who knew I'd eventually pine for `authzQueryConds`?)

I have a partial branch for the second one, but I'm a bit stuck on the permission side of things, and haven't had a chance to loop back yet.

### PR DESCRIPTION
**Pull request by LawnGnome**
This addresses the other significant hotspot in #37246 — namely, that retrieving the code hosts for a batch spec (and, in turn, a batch change) was slow.

This turns out to be because we hydrated every changeset spec attached to the batch spec. The actual database query isn't so bad there, but the real cost is in unmarshalling the JSON blob out of the `spec` field into the `ChangesetSpec`. It turns out we don't actually need that, since the only reason we ever hydrated the changeset specs was to access their repo IDs to pass into another query. 🤦 

Anyway, this PR adds a new store method that, given a batch spec ID, returns the repo IDs that the user has access to. (Added fun: we probably leaked code hosts the user didn't have access to before this. Cool.) We still feed this into another query, so _in theory_ we could actually make this faster still by merging these into one big CTE thing, but I looked at the code host query this feeds into and remembered how awful it was, so we're going with this for now.

Performance on my 195 changeset batch change that adds just under a million lines to 195 different repos is significantly improved — the (every five second) calls when previewing the batch spec _or_ viewing the actual batch change went from ~4 seconds per call to 30-60 ms. Not a typo.

I also fixed up a typo-ed type name along the way, and opportunistically renamed `ct` to `bt` in the store test file I was already in.

Since #37250 has landed, merging this fixes #37246.

## Possible future work

* Merge those two queries into one big CTE thing, which might actually give you CTE as you bang your head against the screens of SQL involved.
* Modify `ChangesetSpec` to only unmarshal `Spec` on the first request, which I started doing, but then decided I didn't care enough about right now. We should keep an eye on this in case it's a theme that recurs again later, though.

## Test plan

We have decent coverage of this resolver method, and none of those tests had to change, which is usually the sign of a good performance improvement. Beyond that, store tests were added, and some manual testing was performed both in browser and testing the performance of the GraphQL resolvers.

### CODE REVIEW

File: enterprise/cmd/frontend/internal/batches/resolvers/code_host_connection.go [hunk 1]
```diff
@@ -156,19 +156,19 @@ type idType struct {
 	externalServiceType string
 }
 
-type emptyEatchChangesCodeHostConnectionResolver struct {
+type emptyBatchChangesCodeHostConnectionResolver struct {
```
Comment (eseliger) [inline enterprise/cmd/frontend/internal/batches/resolvers/code_host_connection.go:159]: lol

File: enterprise/internal/batches/store/batch_specs.go [hunk 2]
```diff
@@ -464,6 +465,58 @@ ON
...
+SELECT DISTINCT repo.id
+FROM repo
+LEFT JOIN changeset_specs ON repo.id = changeset_specs.repo_id
+LEFT JOIN batch_specs ON changeset_specs.batch_spec_id = batch_specs.id
+WHERE
+	repo.deleted_at IS NULL
+	AND batch_specs.id = %s
+	AND %s -- authz query conds
+ORDER BY
+	repo.id ASC
```
Comment (eseliger) [inline enterprise/internal/batches/store/batch_specs.go:517]: do we actually need those to be in order? Could save some DB effort to order them.

**Review by eseliger — APPROVED** [review]

**Comment by eseliger** [comment]
Agreed, we too often unmarshal tons of JSON we don't actually end up using 😢

**Review by Piszmog — APPROVED** [review]

File: enterprise/internal/batches/store/batch_specs.go (hunk 2 above)
Comment (LawnGnome) [reply enterprise/internal/batches/store/batch_specs.go:517]: Technically, no: the `ORDER BY` is a remnant of an earlier mega-query that went all the way to `external_services`, in which case the order was important for pagination reasons. I'll remove it.

File: enterprise/internal/batches/store/batch_specs_test.go [hunk 3]
```diff
@@ -808,3 +811,72 @@ func TestStoreGetBatchSpecStats(t *testing.T) {
...
+	s := New(db, &observation.TestContext, nil)
+
+	// Create two repos, one of which will be visible to everyone, and one which
+	// won't be.
+	globalRepo, _ := bt.CreateTestRepo(t, ctx, db)
+	hiddenRepo, _ := bt.CreateTestRepo(t, ctx, db)
+
+	// One, two princes kneel before you...
+	//
+	// That is, we need an admin user and a regular one.
```
Comment (courier-new) [inline enterprise/internal/batches/store/batch_specs_test.go:828]: 😂❤️

**Review by courier-new — APPROVED** [review]

**Pull request merged**